File: clients/windows/Dayflow.Windows/Core/DayflowEventEnvelope.cs
using System.Text.Json.Serialization;

namespace Dayflow.Windows.Core;

public sealed record DayflowEventEnvelope(
    [property: JsonPropertyName("event_id")] string EventId,
    [property: JsonPropertyName("device_id")] string DeviceId,
    [property: JsonPropertyName("logical_clock")] ulong LogicalClock,
    [property: JsonPropertyName("schema_version")] ushort SchemaVersion,
    [property: JsonPropertyName("key_version")] uint KeyVersion,
    [property: JsonPropertyName("nonce")] string Nonce,
    [property: JsonPropertyName("ciphertext")] string Ciphertext)
{
    public const ushort CurrentSchemaVersion = 1;
    public const ulong MaxLogicalClock = 9_007_199_254_740_991;

    public static bool HasValidEncryptedFieldShape(string nonce, string ciphertext)
    {
        return TryDecodeWireBase64(nonce, out var nonceBytes)
            && TryDecodeWireBase64(ciphertext, out var ciphertextBytes)
            && nonceBytes.Length == 24
            && ciphertextBytes.Length >= 16;
    }

    private static bool TryDecodeWireBase64(string value, out byte[] bytes)
    {
        bytes = Array.Empty<byte>();
        if (string.IsNullOrEmpty(value)) return false;

        foreach (var character in value)
        {
            if (!((character >= 'A' && character <= 'Z')
                || (character >= 'a' && character <= 'z')
                || (character >= '0' && character <= '9')
                || character is '+' or '/' or '-' or '_' or '='))
            {
                return false;
            }
        }

        var firstPadding = value.IndexOf('=');
        if (firstPadding >= 0)
        {
            if (value.Length % 4 != 0 || firstPadding < value.Length - 2) return false;
            for (var index = firstPadding; index < value.Length; index++)
            {
                if (value[index] != '=') return false;
            }
        }

        var unpadded = firstPadding >= 0 ? value[..firstPadding] : value;
        if (unpadded.Length % 4 == 1) return false;
        var normalized = unpadded.Replace('-', '+').Replace('_', '/');
        normalized += new string('=', (4 - normalized.Length % 4) % 4);
        try
        {
            bytes = Convert.FromBase64String(normalized);
            return true;
        }
        catch (FormatException)
        {
            return false;
        }
    }
}

File: clients/android/app/src/main/java/app/dayflow/android/DayflowEventEnvelope.kt
package app.dayflow.android

import java.util.Base64

data class DayflowEventEnvelope(
    val eventId: String,
    val deviceId: String,
    val logicalClock: ULong,
    val schemaVersion: UShort,
    val keyVersion: UInt,
    val nonce: String,
    val ciphertext: String,
) {
    companion object {
        const val CURRENT_SCHEMA_VERSION = 1
        const val MAX_LOGICAL_CLOCK = 9_007_199_254_740_991L
    }
}

/**
 * The Rust core and relay accept padded/unpadded standard or URL-safe base64.
 * Keep the native persistence boundary just as strict: a valid XChaCha20
 * nonce is 24 bytes and the ciphertext must include its 16-byte auth tag.
 */
internal object DayflowWireEnvelopeValidation {
    private const val EVENT_NONCE_BYTES = 24
    private const val EVENT_AUTH_TAG_BYTES = 16

    fun hasValidEncryptedFieldShape(nonce: String, ciphertext: String): Boolean =
        decode(nonce)?.size == EVENT_NONCE_BYTES &&
            decode(ciphertext)?.size?.let { it >= EVENT_AUTH_TAG_BYTES } == true

    fun decode(value: String): ByteArray? {
        if (value.isEmpty() || value.any { it !in "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=_-" }) {
            return null
        }

        val firstPadding = value.indexOf('=')
        if (firstPadding >= 0) {
            if (value.length % 4 != 0 || firstPadding < value.length - 2 ||
                value.substring(firstPadding).any { it != '=' }
            ) {
                return null
            }
        }

        val unpadded = value.substringBefore('=').replace('-', '+').replace('_', '/')
        if (unpadded.length % 4 == 1) return null
        val padded = unpadded + "=".repeat((4 - unpadded.length % 4) % 4)
        return runCatching { Base64.getDecoder().decode(padded) }.getOrNull()
    }
}
